source/common/init/watcher_impl.h

```diff
@@ -51,23 +54,86 @@ class WatcherHandleImpl : public WatcherHandle, Logger::Loggable<Logger::Id::ini
 
-  // The callback function, called via WatcherHandleImpl by either the target or the manager
+  // The callback function, called via WatcherHandleImpl by either the target or the manager.
   const std::shared_ptr<ReadyFn> fn_;
 };
 
+/**
+ * A WatcherHandleImpl functions as a weak reference to a Watcher. It is how a TargetImpl safely
+ * notifies a ManagerImpl that it has initialized, and likewise it's how ManagerImpl safely tells
+ * its client that all registered targets have initialized, with no guarantees about the lifetimes
+ * of the manager or client.
+ *
+ * We restrict the watcher_ inside ManagerImpl to be constructed with the 'TargetAwareReadyFn' fn so
+ * that the init manager will get target name information when the watcher_ calls
+ * 'onTargetSendName(target_name)' For any other purpose, a watcher can be constructed with either
+ * Ctor. If you do not need a watcher to carry any string information such as the target_name, the
+ * first type with 'ReadyFn' fn is enough.
+ */
+class TargetAwareWatcherHandleImpl : public WatcherHandle, Logger::Loggable<Logger::Id::init> {
+private:
+  friend class TargetAwareWatcherImpl;
+
+  // Ctor with std::function<void(absl::string_view)> callback function.
+  TargetAwareWatcherHandleImpl(absl::string_view handle_name, absl::string_view name,
+                               std::weak_ptr<TargetAwareReadyFn> fn);
+
+public:
+  // Init::WatcherHandle.
+  bool ready() const override;
+
+private:
+  // Name of the handle (either the name of the target calling the manager, or the name of the
+  // manager calling the client).
+  const std::string handle_name_;
+
+  // Name of the watcher (either the name of the manager, or the name of the client).
+  const std::string name_;
+
+  // The watcher's callback function, only called if the weak pointer can be "locked".
+  const std::weak_ptr<TargetAwareReadyFn> fn_;
+};
+
+/**
+ * A TargetAwareWatcherImpl is a WatcherImpl which is specially designed for init manager's internal
+ * watcher_. This watcher_ will monitor all the targets this init manager has added. The callback
+ * function fn has a absl::string_view parameter to pass the target name to init manager.
+ */
+class TargetAwareWatcherImpl : public Watcher, Logger::Loggable<Logger::Id::init> {
+public:
+  /**
+   * @param name a human-readable watcher name, for logging / debugging.
+   * @param fn a callback function to invoke when `ready` is called on the handle.
+   */
+  TargetAwareWatcherImpl(absl::string_view name, TargetAwareReadyFn fn);
+  ~TargetAwareWatcherImpl() override;
+
+  // Init::Watcher.
+  absl::string_view name() const override;
+  WatcherHandlePtr createHandle(absl::string_view handle_name) const override;
+
+private:
+  // Human-readable name for logging.
+  const std::string name_;
+
+  // The callback function with target_name parameter, called via WatcherHandleImpl by either the
+  // target or the manager.
+  const std::shared_ptr<TargetAwareReadyFn> fn_;
+};
+
```

Comment:
Why do you still need these classes? `WatcherImpl`/`WatcherHandleImpl` can be just like these with default empty names?

Reply:
Sorry, do you mean that change ReadyFn to ```std::function<void(absl::string_view)>```?

Reply:
Yes, why not?

Reply:
The problem is, in the current code base, most watchers are defined as ```[]() { }``` and the absl::string_view parameter will not automatically transform an empty parameter into a empty ```string_view```. I have to change all those definitions from ```[](){} ```into ```[](""){}```

Reply:
You can overload constructor to take [](){} and wrap with empty string?

Reply:
Several attempts I have tried on this project:
1. Directly change ```ReadyFn``` from ```std::function<void()>``` into  ```std::function<void(absl::string_view)>```. Turns out that due to the most definitions in the current code base are ```[](){}```(Like hundreds of definitions in tests and other places). There will be errors unless I change all of them from ```[](){}``` into ```[](""){}```. This solution is not good because there are too many current definitions to update and for potential wrong definition in the future usage.
2. Add additional Ctor, which could receive both ```ReadyFn```  and ```std::function<void(string_view)>```. Any one of them could work. ReadyFn works well for existing definitions and the new one support the new feature(the internal watcher of manager).

Reply:
> Yes, why not?

Since we need to send name to manager, the ```onTargetReady``` callback has to receive a string parameter now. So the callback passed to watcher has to be called somewhere with a string name. I thought about template and lambda, but that will not change the truth that watcher needs to pass string and call ```onTargetReady(string)```.

Reply:
> You can overload constructor to take {} and wrap with empty string?

I am not sure if that could work as I need to make the function shared. But I will try.

Reply:
> You can overload constructor to take {} and wrap with empty string?

I think the two types of Ctor in my previous version is what is done here. Basically, the wrapping looks like ```fn_( [](""){std::move(fn); ) )```. I don't see the difference of using two Ctors and wrapping one into the other.

Reply:
Lets assume ```A=std::function<void()>``` and ```B=std::function<void(absl::string_view)>```
Existing definition of watcher receives: ```[](){}``` as function. This is class A. It is able to wrap A into B:
````
B fn_;
Ctor(A) {
    fn_ = []("") { A }
}.
````
But finally, watchers Ctor with A naturally does not need to pass string into A, which means this watcher basically is higher level watcher which does not belong to any manager and it does not need to send name. It's meaningless wrapping its A into B. 
So I suppose it is more clear to make them different(with 2 Ctors or 2 separate Classes).

Reply:
> You can overload constructor to take {} and wrap with empty string?

Given that watcher is now defined as 
```c++
watcher(name, [](string name) {
    onTargetReady(name);
}
``` 
I think its impossible to roll this ```[](string){ f(string); }``` into a ```<void()>```.